compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -394,6 +394,56 @@ public #{declName}(com.microsoft.bond.StructBondType<#{declName}<#{paramList}>>
             else mempty
 
 
+-- builds text for Object.equals(Object) override
+object_equals :: String -> [Field] -> Maybe Type -> Text
+object_equals declName fields structBase = [lt|@Override
+    public boolean equals(Object o) {
+        if (!(o instanceof #{declName})) return false;
+        #{compareBase structBase}
+        final #{declName} other = (#{declName}) o;
+        #{newlineSep 2 compareField fields}
+        return true;
+    }
+|]
+    where
+        compareBase (Just _)    = [lt|if (!(super.equals(o))) return false;|]
+        compareBase Nothing     = mempty
+        compareField Field {..} = [lt|if (!(#{equalsMember fieldType fieldName})) return false;|]
+        equalsMember t f
+            | isJavaPrimitive t = [lt|this.#{f} == other.#{f}|]
+            | otherwise         = [lt|(this.#{f} == null && other.#{f} == null) || (this.#{f} != null && this.#{f}.equals(other.#{f}))|]
+
+
+-- builds text for Object.hashCode() override
+object_hashCode :: [Field] -> Maybe Type -> Text
+object_hashCode fields structBase = [lt|@Override
+    public int hashCode() {
+        int result = 17;
+        #{newlineSep 2 hash hashInputs}
+        return result;
+    }
+|]
+    where
+        hash codeExpr = [lt|result += #{codeExpr};
+        result *= 0xeadbeef;
+        result ^= result >> 16;|]
+
+        hashInputs = (hashBase structBase) ++ hashFields
+
+        hashBase (Just t)  = [[lt|super.hashCode()|]]
+        hashBase Nothing = []
+
+        hashFields = map (\Field {..} -> hashCode fieldType fieldName) fields
+        hashCode BT_Bool e    = [lt|(#{e} ? 0 : 1)|]
+        hashCode BT_Int64 e   = [lt|(#{e} ^ (#{e} >> 32)) & 0xFFFFFFFF|]
+        hashCode BT_UInt64 e  = [lt|(#{e} ^ (#{e} >> 32)) & 0xFFFFFFFF|]
+        hashCode BT_Float e   = [lt|Float.floatToRawIntBits(#{e})|]
+        hashCode BT_Double e  = [lt|(Double.doubleToRawLongBits(#{e}) ^ (Double.doubleToRawLongBits(#{e}) >> 32)) & 0xFFFFFFFF|]
```

Comment:
>Double.doubleToRawLongBits(#{e}) ^ (Double.doubleToRawLongBits(#{e}) >> 32) [](start = 37, length = 75)

This won't work well if comparing 0 with -0. Equality holds but hash codes are different.